lib/Memoizer.js

```diff
@@ -1,61 +1,117 @@
-/* globals Services */
+/* globals Services, Task */
 
 "use strict";
 const {Cu} = require("chrome");
 const simplePrefs = require("sdk/simple-prefs");
+const ss = require("sdk/simple-storage");
 Cu.import("resource://gre/modules/Services.jsm");
+Cu.import("resource://gre/modules/Task.jsm");
+
+const concat = String.prototype.concat;
 
 function Memoizer() {
   this._cacheEnabled = simplePrefs.prefs["query.cache"];
-  this._memo_all = {};
+
+  if (!ss.storage.queryCache) {
+    ss.storage.queryCache = {};
+  }
   this._onPrefChange = this._onPrefChange.bind(this);
   simplePrefs.on("", this._onPrefChange);
 }
 
 Memoizer.prototype = {
 
+  /**
+   * Return a function that will wrap a given function and cache its results.
+   *
+   * @param {String} key
+   *   a key to store/retrieve cached values
+   *
+   * @param {Function} func
+   *   a function to wrap for caching
+   *
+   * @return {Promise}
+   *   A promise whose result contains data
+   */
   memoize(key, func) {
-    let concat = String.prototype.concat;
     return (...args) => {
-      if (!this._memo_all[key]) {
-        this._memo_all[key] = {};
-      }
-      let memo = this._memo_all[key];
-
       // allow object params to be passed, as is expected of query functions
-      let parts = [];
-      for (let arg of args) {
-        if (typeof arg === "object") {
-          // sort so we have a stable key
-          let props = Object.keys(arg).sort();
-          for (let prop of props) {
-            parts.push(`${prop}:${arg[prop]}`);
-          }
-        } else {
-          parts.push(arg);
-        }
-      }
-      let repr = concat.call(parts);
+      let repr = this._makeRepr(args);
 
-      if (this._cacheEnabled && repr in memo) {
+      if (this._cacheEnabled && this.hasRepr(key, repr)) {
         Services.obs.notifyObservers(null, `${key}-cache`, "hit");
-        return memo[repr];
+        return new Promise(resolve => resolve(this.getData(key, repr)));
       } else {
         Services.obs.notifyObservers(null, `${key}-cache`, "miss");
```

Comment:
Yeah, the Task is there to execute an async query if need be.
I.e. it will always execute async.

You are right, it might make data non-straightforward if we are counting how often there are cache misses. We should still obtain information on relative changes in this number though.

That said, we're not counting cache misses yet. I think we should tackle that when we do.